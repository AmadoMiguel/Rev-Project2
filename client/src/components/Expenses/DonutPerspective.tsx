import React, { Fragment } from 'react';
import DonutGraph from '../data/DonutGraph';
import NewExpense from './NewExpenseDialog';
import { FormControlLabel, Checkbox } from '@material-ui/core';

function DonutPerspective(props:any) {
    return (
        <Fragment>
            {((props.expenses.length == 1 && props.expenses[0].id != 0)
            || props.expenses.length > 1) &&
            <div>
                <h3>{props.showMonthly ? "This month" : "Overall"} expenses:
                {props.loading ? "..." : props.showMonthly ? " $" + props.monthExpTotal : 
                " $" + props.expTotal}</h3>
                <i style={{ color: 'grey', fontSize: '14px' }}>
                Click on any section of the graphic to view details.</i>
                <DonutGraph
                data={props.graphData}
                labels={props.labels}
                important='Emergency'
                isMobileView={props.view}
                handleElementClick={props.selectedType} />
                <NewExpense
                types={props.expTypes}
                createExpense={props.createExpense}
                view={props.view} />
                {/* Toggles between view monthly expenses and overall expenses */}
                <FormControlLabel
                control= {
                    <Checkbox
                    checked={props.showMonthly}
                    onChange={props.showMonthly ? () => props.viewMonthlyExpenses(false) :
                        () => props.viewMonthlyExpenses(true)}
                    value="checkedB"
                    color="primary" />
                }
                style= {{ marginLeft: '5px' }}
                label= "This month" />
            </div>
            }
        </Fragment>
    )
}

export default DonutPerspective;