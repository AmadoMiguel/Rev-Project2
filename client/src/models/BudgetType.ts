export interface BudgetType {
    id: number,
    type: string
}