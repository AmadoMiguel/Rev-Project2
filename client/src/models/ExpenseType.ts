export interface ExpenseType {
    id: number,
    type: string
}