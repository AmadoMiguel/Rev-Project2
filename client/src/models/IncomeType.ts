export interface IncomeType {
    id: number,
    type: string
}