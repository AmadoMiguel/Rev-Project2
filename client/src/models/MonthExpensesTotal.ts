export interface MonthExpensesTotal {
    month: string,
    total: number
}