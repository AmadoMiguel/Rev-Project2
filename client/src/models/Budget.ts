import { BudgetType } from "./BudgetType";

export interface Budget {
    id: number,
    userId: number,
    amount: number,
    budgetType:BudgetType
}