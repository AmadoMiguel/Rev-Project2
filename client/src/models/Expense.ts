import { User } from "./User";
import { ExpenseType } from "./ExpenseType";

export interface Expense {
    id: number,
    userId: number,
    userInfo: User,
    expenseType: ExpenseType,
    date: Date,
    description: string,
    amount: number
}