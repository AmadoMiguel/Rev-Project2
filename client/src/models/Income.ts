import { IncomeType } from "./IncomeType";

export interface Income {
    id: number, 
    userId: number,
    incomeType: IncomeType
    description: string,
    amount: number
}