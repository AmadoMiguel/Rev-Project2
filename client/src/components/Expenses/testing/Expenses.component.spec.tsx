import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { configure, mount, shallow, ReactWrapper } from 'enzyme';
import axios from 'axios';
import ExpensesComponent, { IExpenseProps } from '../Expenses.component';
import { User } from '../../../models/User';
import { Expense } from '../../../models/Expense';
import DonutPerspective from '../DonutPerspective';
import { Paper } from '@material-ui/core';
import configureStore from 'redux-mock-store';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import { Provider } from 'react-redux';
import promiseMiddleware from 'redux-promise-middleware';
import reduxThunk from 'redux-thunk';
import { BarLoader } from 'react-spinners';

// Mock all required properties (redux) and axios calls
// Mock the redux store with required middleware
const createStore = configureStore([reduxThunk]);
const userInfo:User = {
        id: 2, 
        firstName: "Post",
        lastName: "Man",
        email: "postman@mail.com",
        username: "postman",
        token: "ThisIsATokenMock"
}
const props =  {
    ui:{
        isMobileView: false
    },
    history:{},
    user : {
        userInfo: userInfo,
        isLoggedIn: true
    },
    userExpenses: {
        expenseTypes:[
            {id: 1,
            type: "Bills"},
            {id: 2,
            type: "Food"},
            {id: 3,
            type: "Emergency"},
            {id: 4, 
            type: "Entertainment"},
            {id: 5,
            type: "Other"}
        ],
        expenses: [
            {
                id:1, 
                userId:userInfo.id,
                user:userInfo,
                expenseType:{
                    id: 1,
                    type: "Bills"
                },
                date: new Date("03-09-2019"),
                description: "Phone bill",
                amount: 30
            },
            {
                id:2, 
                userId:userInfo.id,
                user:userInfo,
                expenseType:{
                    id: 2,
                    type: "Food"
                },
                date: new Date("03-10-2019"),
                description: "Last Revature lunch",
                amount: 20
            }
        ],
        expensesTotal:50,
        thisMonthExpenses: [
            {
                id:2, 
                userId:userInfo.id,
                user:userInfo,
                expenseType:{
                    id: 2,
                    type: "Food"
                },
                date: new Date("03-10-2019"),
                description: "Last Revature lunch",
                amount: 20
            }
        ],
        thisMonthExpensesTotal:20,
        thisYearTotalExpensesByMonth: [
            {
                month: "September",
                total: 30
            },
            {
                month: "October",
                total: 20
            }
        ]
    },
    setExpenses: jest.fn().mockImplementation(),
    setExpensesTotal: jest.fn().mockImplementation(),
    setThisMonthExpenses: jest.fn().mockImplementation(),
    setThisMonthExpensesTotal: jest.fn().mockImplementation(),
    setThisYearExpensesTotalByMonth: jest.fn().mockImplementation()
}
// Create a new fake expense in order to mock the axios post (create) method
const newFakeExpense:Expense = {
    id: 3,
    userId: props.user.userInfo.id,
    user: props.user.userInfo,
    expenseType: {
        id: 4,
        type: "Entertainment"
    },
    date: new Date(),
    description: "Testing",
    amount: 100
}


configure({adapter: new Adapter()});
describe("Testing <Expenses />", () => {
    describe("initial rendering", () => {
        it("bar loader component should render first on expenses component", () => {
            // Mock the expenses component
            let setIsLoading:(state:boolean) => void;
            const useEffectSpy = jest.spyOn(React, "useEffect");
            useEffectSpy.mockImplementationOnce(() => {
                setIsLoading(false);
            });
            const store = createStore(props);
            const expensesComponentRenderMock = renderer.create(
                <Provider store = {store}>
                    <ExpensesComponent {...props}/>
                </Provider>
            );
            const barLoaderMock = expensesComponentRenderMock.root
            .findByType("div").findByType(Paper).findByType("div").findByType(DonutPerspective);
            expect(barLoaderMock).toBeDefined();
        });
    });
    // describe("redux actions for expenses total should be called properly", () => {
        

    // });
});