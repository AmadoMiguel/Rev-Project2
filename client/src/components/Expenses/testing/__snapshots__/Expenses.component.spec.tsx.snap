// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Expenses /> initial rendering the expenses component should be rendered with the redux config 1`] = `
<div
  style={
    Object {
      "textAlign": "center",
    }
  }
>
  <h2
    style={
      Object {
        "color": "#FFFFFF",
      }
    }
  >
    Your expenses
  </h2>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    style={
      Object {
        "backgroundColor": "rgba(220,245,230,0.9)",
        "height": "60%",
        "margin": "5px auto",
        "padding": "10px",
        "width": "50%",
      }
    }
  >
    <div
      style={
        Object {
          "display": "inline-block",
          "margin": "150px",
        }
      }
    >
      <div
        className="css-14q3vga-wrapper-Loader"
      >
        <div
          className="css-x54n5o-style"
        />
        <div
          className="css-19d6xdh-style"
        />
      </div>
    </div>
  </div>
</div>
`;
